packages/utils/src/internals/error_snapshotter.ts

```diff
@@ -0,0 +1,100 @@
+import crypto from 'node:crypto';
+
+import type { CrawlingContext } from '@crawlee/basic';
+import type { KeyValueStore } from '@crawlee/core';
+import type { Page as PlaywrightPage } from 'playwright';
+import type { Page as PuppeteerPage } from 'puppeteer';
+
+import type { ErrnoException } from './error_tracker';
+
+/**
+ * ErrorSnapshotter class is used to capture a screenshot of the page and a snapshot of the HTML when an error occur during web crawling.
+ */
+export class ErrorSnapshotter {
+    static readonly MAX_ERROR_CHARACTERS = 30;
+    static readonly BASE_MESSAGE = 'An error occurred';
+    static readonly SNAPSHOT_PREFIX = 'SNAPSHOT';
+    static readonly KEY_VALUE_STORE_PATH = 'https://api.apify.com/v2/key-value-stores';
+
+    /**
+     * Capture a snapshot of the error context.
+     */
+    async captureSnapshot(error: ErrnoException, context: CrawlingContext): Promise<{ screenshotFilename?: string; htmlFilename?: string }> {
+        const page = context?.page as PuppeteerPage | PlaywrightPage | undefined;
+        const body = context?.body;
+
+        const keyValueStore = await context?.getKeyValueStore();
+        // If the key-value store is not available, or the body and page are not available, return empty filenames
+        if (!keyValueStore || (!body && !page)) {
+            return {};
+        }
+
+        const filename = this.generateFilename(error);
+        const screenshotFilename = page ? await this.captureScreenshot(page, keyValueStore, filename) : undefined;
+
+        let htmlFilename: string | undefined;
+        if (typeof body === 'string') {
+            htmlFilename = await this.saveHTMLSnapshot(body, keyValueStore, filename);
+        } else if (page) {
+            const html = await page?.content() || '';
+            htmlFilename = html ? await this.saveHTMLSnapshot(html, keyValueStore, filename) : '';
+        }
+
+        const basePath = `${ErrorSnapshotter.KEY_VALUE_STORE_PATH}/${keyValueStore.id}/records`;
+
+        return {
+            screenshotFilename: screenshotFilename ? `${basePath}/${screenshotFilename}` : undefined,
+            htmlFilename: htmlFilename ? `${basePath}/${htmlFilename}` : undefined,
+        };
+    }
+
+    /**
+     * Capture a screenshot of the page (For Browser only), and return the filename with the extension.
+     */
+    async captureScreenshot(page: PuppeteerPage | PlaywrightPage, keyValueStore: KeyValueStore, filename: string): Promise<string | undefined> {
+        try {
+            const screenshotBuffer = await page.screenshot({ fullPage: true, type: 'png' });
```

Comment:
What is the reason to not use [utils.saveSnapshot](https://github.com/apify/crawlee/blob/a8cd603611b9804e49eef9744ae02173774e6a6a/packages/playwright-crawler/src/internals/utils/playwright-utils.ts#L514)? It is in context for both Puppeteer and Playwright.

If we don't want to use it, i would at least use the default quality and jpeg so it is not as big

Reply:
No reason, but after testing it I noticed that it saves the file name twice, I add a commit to fix it with the current pull request

![image](https://github.com/apify/crawlee/assets/36267291/8e076fb4-4044-49f2-a8ff-d4668b8f5f28)

Reply:
Th double extension is because locally we add the extension while on platform we don't. Not sure how to have it nicer

Reply:
@metalwarrior665 This is the code for adding the extension for the file, I updated it to check if we are on the platform or not, would this solve the issue?

```
const { APIFY_IS_AT_HOME } = process.env;

if (saveScreenshot) {
    const screenshotName = APIFY_IS_AT_HOME ? `${key}.jpg` : key;
    const screenshotBuffer = await page.screenshot({ fullPage: true, quality: screenshotQuality, type: 'jpeg', animations: 'disabled' });
    await store.setValue(screenshotName, screenshotBuffer, { contentType: 'image/jpeg' });
}

if (saveHtml) {
    const htmlName = APIFY_IS_AT_HOME ? `${key}.html` : key;
    const html = await page.content();
    await store.setValue(htmlName, html, { contentType: 'text/html' });
}
```

Reply:
@HamzaAlwan I'm not sure if this works since you are storing them as the same key. Not sure at which point the extension is appended, could be a bit unclear and breaking. But if it works, I would probably be for it.

Reply:
I made this update inside `playwright_utils.ts` and `puppeteer_utils.ts` inside the `saveSnapshot` function.

cc @B4nan